### PR TITLE
ASQ-136 - Extend SUSOM Database API to support LocalDate type

### PR DESCRIPTION
**Pull request by glenna-m**
Currently the Susom database API supports the date type as date/time only.  This is really helpful for ensuring fields like create and update times are properly handled, but it's not ideal for tables with fields like date of birth which are used by the sleep survey.

For the ASQ prototype, we should be using a true DB date on some fields.

This ticket tracks the work to extend the API to support this real DB date type.

This has been tested on Derby, HSQL, Postgres (the prototype target DB).  I do not currently have an Oracle DB or SQL Server DB available for testing.  I'm not sure if I need to install those databases on my laptop or if Garrick has a developer test environment that can be used.  Before investing time in that, I wanted to verify whether Garrick is OK with the general approach.

### CODE REVIEW

**Comment by jpallas** [comment]
The thing I would like to see is a comment somewhere (Schema.ColumnType?) that explains the relationship between Date and DbDate.  People who are reading the code will not have the commit message in front of them, and will likely be confused.

File: src/main/java/com/github/susom/database/Flavor.java [hunk 1]
```diff
@@ -311,6 +319,11 @@ public String typeDate() {
       return "timestamp(3)";
     }
 
+    @Override
+    public String typeDbDate() {
+      return "date";
```
Comment (garricko) [inline src/main/java/com/github/susom/database/Flavor.java]: Note that for Oracle, DATE includes time to second precision, and adjusts for daylight savings, which can sometimes shift things to the wrong day. Other databases might also do strange things, which is partly why I hadn't tried to figure out local dates yet... :)

File: src/main/java/com/github/susom/database/StatementAdaptor.java [hunk 2]
```diff
@@ -182,6 +186,15 @@ public Object nullDate(Date arg) {
     return new Timestamp(arg.getTime());
   }
 
+  public Object nullDbDate(LocalDate arg) {
+    if (arg == null) {
+      return new SqlNull(Types.DATE);
+    }
+    Logger log = LoggerFactory.getLogger("StatementAdapter.class");
+    log.info("nullDbDate converting arg"+arg.toString()+ " to " + java.sql.Date.valueOf(arg));
```
Comment (garricko) [inline src/main/java/com/github/susom/database/StatementAdaptor.java]: Sure you want to be logging here?

File: src/test/java/com/github/susom/database/test/CommonTest.java [hunk 3]
```diff
@@ -1503,19 +1563,60 @@ public Date process(Rows rs) throws Exception {
...
+            .argDbDate(dateOfBirth)
+            .insert(1);
+
+    LocalDate testDate = db.toSelect("select d from dbtest").queryDbDateOrNull();
+    assertEquals(dateOfBirth, testDate);
+  }
+
+
+  @Test
+  public void dbDateRoundTrip() {
```
Comment (garricko) [inline src/test/java/com/github/susom/database/test/CommonTest.java]: The thing I really worry about with these, which I haven't figured out how to test well, is combinations of JVM and database time zones, and edge cases like daylight savings shifts. Especially when you are trying to store a calendar day in a database column that is actually a point in time (e.g. Oracle DATE), things go spectacularly wrong in rare circumstances.

I have seen a couple strategies to help mitigate those, each of which has its own problems: a) use a column the database will definitely not treat as a datetime (e.g. integer for Julian days, or string like yyyymmdd), or b) mandate the database be in GMT/UTC and set time component to noon rather than truncating to midnight, so DST or TZ or leap second shifts don't move you to a different calendar day.

File: src/main/java/com/github/susom/database/StatementAdaptor.java [hunk 4]
```diff
@@ -80,6 +82,8 @@ public void addParameters(PreparedStatement ps, Object[] parameters) throws SQLE
         } else {
           ps.setNull(i + 1, sqlNull.getType());
         }
+      } else if (parameter instanceof java.sql.Date) {
+        ps.setDate( i + 1, (java.sql.Date) parameter);
```
Comment (garricko) [inline src/main/java/com/github/susom/database/StatementAdaptor.java]: See longer comments in CommonTest. Probably requires lots of testing, but at least in some cases like Oracle, I would be tempted to explicitly set the time component to noon to make accidental date shifts less likely.

File: src/main/java/com/github/susom/database/StatementAdaptor.java (hunk 2 above)
Comment (glenna-m) [reply src/main/java/com/github/susom/database/StatementAdaptor.java]: I didn't.  I caught that and fixed it in the new commit, before reading this .

File: src/main/java/com/github/susom/database/Flavor.java [hunk 5]
```diff
@@ -83,6 +83,9 @@ public String typeDate() {
       return "timestamp";
     }
 
+    @Override
+    public String typeDbDate() { return "date"; }
```
[the remark this answers is not in the document]
Comment (glenna-m) [reply src/main/java/com/github/susom/database/Flavor.java]: OK - I didn't realize you had commented, so will roll this shortly.  I prefer multiline but was emulating what the code was with the equivalent date file.  I'll fix both.

File: src/main/java/com/github/susom/database/StatementAdaptor.java (hunk 4 above)
Comment (glenna-m) [reply src/main/java/com/github/susom/database/StatementAdaptor.java]: Yes - my use case is Postgres, not Oracle, but I recognize we ned to support that.  I knew Oracle had a DateTime type, and did not realize they were stashing timestamps on Date. Thank you for pointing that gotcha out. 

For the sleep survey prototype, the reason I was doing this work, we have generally been using the server time zone for our data.  Do you consider this acceptable for things like DoB, with perhaps an option to override this? 

I will see about modifying the code to force the time to noon for Oracle on insert/update for our dates, and add some testing for it.

I'm not sure how far you want me to take this implementation?  I undertook this mostly so I could get the sleep survey up and running with susom backend standards like vert.x and database.   The noon timestamp should meet our requirements, but that may not be acceptable.  Do we need to talk further design?

File: src/main/java/com/github/susom/database/Flavor.java (hunk 1 above)
Comment (jpallas) [reply src/main/java/com/github/susom/database/Flavor.java]: DATE, TIME WITH TIMEZONE, TIMESTAMP WITH TIMEZONE, TIME WITHOUT TIMEZONE, TIMESTAMP WITHOUT TIMEZONE, this is all perfectly clear in the SQL standard.  And it's not as if any of the major databases diverge from the standard.  I don't see the problem. 🤣

File: src/main/java/com/github/susom/database/StatementAdaptor.java (hunk 4 above)
Comment (glenna-m) [reply src/main/java/com/github/susom/database/StatementAdaptor.java]: BTW, I cannot set the time to noon with LocalDate - there is no time component.  And LocalDateTime has no notion of Timezone either.  I would have to convert these things to Timestamps to do this.   

I am really sure how helpful this would be.    
Oracle and SQLServer 0 out the time related fields including TimeZone when given a LocalDate.  Postgres actually has a decent Date implementation in their driver.  

The only DB that does not is HSQL.  It 0's out time, but puts a timezone on everything.  As a result, if you change timezone, it fails to match LocalDate/LocalDateTime values.  It's a documented crappy implementation and I don't see a clean way to work around it.  Even if I set time to noon, it will fail to match if the client is in a different timezone.  At one point, I tried setting it to noon with UTC, but it really doesn't help.  Noon UTC means it would change a date like 8/29 to 8/28 on our systems, and then the roundtrip lookup still fails because adding a timezone of PST to 8/28 just gives us an 8/28 date.

**Comment by glenna-m** [comment]
> Looks good.
> 
> Minor nit but maybe "date with timestamp" and "date without timestamp" should be "date with/without time component" or "time part", since timestamp includes a date for both SQL and Java.
> 
> Editorial comment: java.sql is awful. They should have done java.nsql for Java 8 using java.time and streams.

Yes - I agree that java.sql should have been replaced properly.  

I did update my comments as you suggested.

File: src/main/java/com/github/susom/database/Schema.java [hunk 6]
```diff
@@ -140,12 +144,28 @@ public Table addTableFromRow(String tableName, Row r) {
         case Types.NCLOB:
           table.addColumn(names[i]).asClob();
           break;
-        case Types.TIMESTAMP:
-          table.addColumn(names[i]).asDate();
-          break;
+
+        // The date type is used for a true date - no time info.
+        // It must be processed before TimeStamp because dates are also
+        // recognized as timestamp type.
```
[the remark this answers is not in the document]
Comment (glenna-m) [reply src/main/java/com/github/susom/database/Schema.java]: Modified as you suggested

File: src/main/java/com/github/susom/database/StatementAdaptor.java (hunk 4 above)
Comment (glenna-m) [reply src/main/java/com/github/susom/database/StatementAdaptor.java]: BTW, I added a test that tries every timezone in common test.  Current implementation is returning the date I stored, for Derby, Postgres, Oracle, and SQL Server.  Only HSQL fails this.
